src/librustc_mir/interpret/eval_context.rs

```diff
@@ -1314,47 +1316,117 @@ impl<'a, 'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> EvalContext<'a, 'mir, 'tcx, M
+            match value {
+                Scalar::Bits {..} => {
+                    let bits = value.to_bits(size)?;
+                    Ok(bound.contains(&bits))
+                },
+                Scalar::Ptr(_) => {
+                    // we cannot reason about Ptr numeric ranges other than zero
+                    if bound.start_bound() == Bound::Included(&0) && bound.end_bound() == Bound::Included(&0) {
+                        Ok(false)
+                    } else {
+                        Ok(true)
+                    }
+                }
+            }
+        }
+        let (lo, hi) = scalar.valid_range.clone().into_inner();
+        if lo > hi {
+            if in_range(value, size, ..=hi)? || in_range(value, size, lo..)? {
+                Ok(())
+            } else {
+                err!(ValidationFailure(format!(
+                    "encountered {:?} but expected something in the range {:?} or {:?}",
+                    value,
+                    ..=hi,
+                    lo..,
+                )))
+            }
+        } else {
+            if in_range(value, size, scalar.valid_range.clone())? {
+                Ok(())
+            } else {
+                err!(ValidationFailure(format!(
+                    "encountered {:?} but expected something in the range {:?}",
+                    value,
+                    scalar.valid_range,
+                )))
+            }
+        }
+    }
+
     pub fn validate_ptr_target(
         &self,
         ptr: Pointer,
         ptr_align: Align,
-        ty: Ty<'tcx>
+        mut layout: TyLayout<'tcx>,
     ) -> EvalResult<'tcx> {
-        match ty.sty {
-            ty::TyBool => {
-                self.memory.read_scalar(ptr, ptr_align, Size::from_bytes(1))?.to_bool()?;
-            }
-            ty::TyChar => {
-                let c = self.memory.read_scalar(ptr, ptr_align, Size::from_bytes(4))?.to_bits(Size::from_bytes(4))? as u32;
-                match ::std::char::from_u32(c) {
-                    Some(..) => (),
-                    None => return err!(InvalidChar(c as u128)),
+        self.memory.dump_alloc(ptr.alloc_id);
+        trace!("validate_ptr_target: {:#?}", layout);
+        if let layout::Variants::Tagged { .. } = layout.variants {
+            let variant_index = self.read_discriminant_as_variant_index(
+                Place::from_ptr(ptr, ptr_align),
+                layout,
+            )?;
+            layout = layout.for_variant(self, variant_index);
+            trace!("variant layout: {:#?}", layout);
+        }
+        match layout.fields {
+            // primitives are unions with zero fields, because that's totally obvious and documented
+            layout::FieldPlacement::Union(0) => {
+                match layout.abi {
+                    // nothing to do, whatever the pointer points to, it is never going to be read
+                    layout::Abi::Uninhabited => Ok(()),
```

Comment:
this check is only done in this case, is that because it's the (only?) base case of recursive `validate_ptr_target`?

if so is it only the inner-most types that have this property on the other cases? as in, is it valid to recurse down to here on nested types or would they themselves be Uninhabited in a way that would mean they aren't recursable?

Reply:
I'm not sure. I'll investigate. But I also do remember that there were some issues with aggregates containing uninhabited types not being uninhabited, because you could create a local variable of such an aggregate and then initialize the inhabited parts